### PR TITLE
WIP: BXMSDOC-6474 SpringBoot: Configuring Security

### PR DESCRIPTION
**Pull request by emmurphy1**
[BXMSDOC-6474](https://issues.redhat.com/browse/BXMSDOC-6474)

[PAM draft
](http://file.ork.redhat.com/~emmurphy/BXMSDOC-6474-pam/#security)
[DM draft](http://file.ork.redhat.com/~emmurphy/BXMSDOC-6474-dm/#security)

### CODE REVIEW

File: doc-content/enterprise-only/springboot/bus-app-security-preauthenticate-proc.adoc [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+[source, java]
+----
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        //@formatter:off
+
+                http
+                    .anonymous().authenticationFilter(new <CLASS_NAME>()).and() // Override anonymousUser
+                    .cors().and().csrf().disable() // Disable CORs & CSRF filters
```
Comment (afalhambra) [inline doc-content/enterprise-only/springboot/bus-app-security-preauthenticate-proc.adoc]: Please remove comment, this is not needed here
```suggestion
                    .cors().and().csrf().disable()
```

File: doc-content/enterprise-only/springboot/bus-app-security-preauthenticate-proc.adoc [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+                    .authorizeRequests()
+                    .antMatchers("/*").permitAll()
+                    .and().headers().frameOptions().disable(); // Disable X-Frame header for h2-console
```
Comment (afalhambra) [inline doc-content/enterprise-only/springboot/bus-app-security-preauthenticate-proc.adoc]: same here
```suggestion
                    .and().headers().frameOptions().disable();
```

File: doc-content/enterprise-only/springboot/bus-app-rh-sso_proc.adoc [hunk 3]
```diff
@@ -53,7 +55,7 @@ keycloak.public-client=true
 keycloak.principal-attribute=preferred_username
 keycloak.enable-basic-auth=true
 ----
-. Modify the `DefaultWebSecurityConfig.java` file to ensure that Spring Security works correctly with RH SSO:
+. Modify the `business-application-service/src/main/java/com/company/service/DefaultWebSecurityConfig.java` file to ensure that Spring Security works correctly with RHSSO:
 +
 [source]
```
Comment (afalhambra) [inline doc-content/enterprise-only/springboot/bus-app-rh-sso_proc.adoc]: This is java source code
```suggestion
[source, java]
```

**Review by afalhambra — CHANGES_REQUESTED** [review]
Very good work @emmurphy1.

I've left some comments in the PR itself, not a big issue, only minor things.

Another thing I need to mention here is the following:

- CHAPTER 4. CONFIGURING THE BUSINESS APPLICATION WITH RED HAT SINGLE SIGN-ON
- Step 3.
It doesn't work straight away as per instructions. You need to have either that `${version.org.keycloak}` value already defined as a property in your pom.xml file or hardcoded to some keycloak version like `3.4.3.Final`

File: doc-content/enterprise-only/springboot/bus-app-security-authorization-proc.adoc [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
++
+* To authorize for a single role, edit the `antMatchers` method as shown in the following example, where `<role>` is the role that that the user must have for access:
++
+[source,java]
+----
+@Configuration("kieServerSecurity")
+@EnableWebSecurity
+public class DefaultWebSecurityConfig extends WebSecurityConfigurerAdapter {
+
+    @Override  // (1)
```
[the remark this answers is not in the document]
Comment (emmurphy1) [reply doc-content/enterprise-only/springboot/bus-app-security-authorization-proc.adoc]: Yes, I didn't see that comment in the code. Thanks.

File: doc-content/enterprise-only/springboot/bus-app-security-authorization-proc.adoc [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+----
+* To authorize a user that has one of a range of roles, edit the `antMatchers` method as shown in the following example, where `<role>` and `<role1>` are each roles the user can have for access:
++
+[source,java]
+----
+@Configuration("kieServerSecurity")
+@EnableWebSecurity
+public class DefaultWebSecurityConfig extends WebSecurityConfigurerAdapter {
+
+    @Override  // (1)
```
[the remark this answers is not in the document]
Comment (emmurphy1) [reply doc-content/enterprise-only/springboot/bus-app-security-authorization-proc.adoc]: Fixed

File: doc-content/enterprise-only/springboot/bus-app-security-disabling-proc.adoc [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+.Procedure
+. In the directory that contains your {PRODUCT} Spring Boot application, open the `business-application-service/src/main/java/com/company/service/DefaultWebSecurityConfig.java` file in a text editor or IDE.
+. Edit the `.antMatchers` method as shown in the following example:
++
+[source, java]
+----
+    @Override
+    protected void configure(HttpSecurity http) throws Exception {
+        //@formatter:off
```
[the remark this answers is not in the document]
Comment (emmurphy1) [reply doc-content/enterprise-only/springboot/bus-app-security-disabling-proc.adoc]: Removed

File: doc-content/enterprise-only/springboot/bus-app-security-disabling-proc.adoc [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+                http
+                    .cors().and().csrf().disable()
+                    .authorizeRequests()
+                    .antMatchers("/*")
+                        .permitAll()
+                    .and().headers().frameOptions().disable();
+
+        // @formatter:on
```
[the remark this answers is not in the document]
Comment (emmurphy1) [reply doc-content/enterprise-only/springboot/bus-app-security-disabling-proc.adoc]: Removed

File: doc-content/enterprise-only/springboot/bus-app-security-preauthenticate-proc.adoc [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+The easiest way to implement preauthentication is to create a custom filter servlet and add it before the security FilterChain in the `DefaultWebSecurityConfig` class. This way, you can inject a customized, profile-based security context, control its contents, and keep it simple.
+
+
+.Prerequisites
+* You have a {PRODUCT} Spring Boot application and you have disabled Spring Security as xref:bus-app-security-disabling-proc_{context}[].
+
+.Procedure
+. Create the following class that extends the `AnonymousAuthenticationFilter` class:
++
+[source]
```
[the remark this answers is not in the document]
Comment (emmurphy1) [reply doc-content/enterprise-only/springboot/bus-app-security-preauthenticate-proc.adoc]: Fixed

File: doc-content/enterprise-only/springboot/bus-app-rh-sso_proc.adoc [hunk 9]
```diff
@@ -1,23 +1,25 @@
...
+* You have a {PRODUCT} Spring Boot application that you created using the http://start.jbpm.org[business applications] website.
 
 .Procedure
-. Download and install RH SSO. For instructions, see the https://access.redhat.com/documentation/en-us/red_hat_single_sign-on/{RH-SSO_VERSION}/html/getting_started_guide/[_Red Hat Single Sign-On Getting Started Guide_].
-. Configure RH SSO:
+. Download and install RHSSO. For instructions, see the https://access.redhat.com/documentation/en-us/red_hat_single_sign-on/{RH-SSO_VERSION}/html/getting_started_guide/[_Red Hat Single Sign-On Getting Started Guide_].
+. Configure RHSSO:
 .. Either use the default master realm or create a new realm.
++
+A Realm manages a set of users, credentials, roles, and groups. A user belongs to and logs into a realm. Realms are isolated from one another and can only manage and authenticate the users that they control.
```
Comment (michelehaglund) [inline doc-content/enterprise-only/springboot/bus-app-rh-sso_proc.adoc]: Should the 1st instance of "Realm" be lower case?

File: doc-content/enterprise-only/springboot/bus-app-rh-sso_proc.adoc [hunk 10]
```diff
@@ -1,23 +1,25 @@
...
 .. Create the `springboot-app` client and set the `AccessType` to public.
 .. Set a valid redirect URI and web origin according to your local setup, for example:
 +
 * Valid redirect URIs: `\http://localhost:8090/*`
 * Web origin: `\http://localhost:8090`
 .. Create realm roles that are used in the application.
 .. Create users that are used in the application and assign roles to them.
-. Add the following dependencies to the service project `pom.xml` file:
+. Add the following dependencies to the Spring Boot project `pom.xml` file:
```
Comment (michelehaglund) [inline doc-content/enterprise-only/springboot/bus-app-rh-sso_proc.adoc:22]: project's?

Add the following dependencies to the Spring Boot project's `pom.xml` file:

File: doc-content/enterprise-only/springboot/bus-app-security-disabling-proc.adoc [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+= Disabling Spring Security in a {PRODUCT} business application
+You can configure Spring Security in a {PRODUCT} business application to provide the security context without authentication.
+//Why would I want to do this?
+
+.Prerequisites
+* You have a {PRODUCT} Spring Boot application.
+
+.Procedure
+. In the directory that contains your {PRODUCT} Spring Boot application, open the `business-application-service/src/main/java/com/company/service/DefaultWebSecurityConfig.java` file in a text editor or IDE.
```
Comment (michelehaglund) [inline doc-content/enterprise-only/springboot/bus-app-security-disabling-proc.adoc]: If this is the 1st instance of "IDE", please write it out: integrated development environment (IDE)

File: doc-content/enterprise-only/springboot/bus-app-security-disabling-proc.adoc [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+                        .permitAll()
+                    .and().headers().frameOptions().disable();
+
+    }
+----
++
+The `PermitAll` method allows any and all requests for the specified URL pattern.
+
+NOTE: Because no security context is passed in the `HttpServletRequest`, Spring  creates an `AnonymousAuthenticationToken`
+and populates the `SecurityContext` with the `anonymousUser` user with no designated roles other than the `ROLE_ANONYMOUS` role. The user will not have access to many of the features of the application, for example they will be unable to action group assigned tasks.
```
Comment (michelehaglund) [inline doc-content/enterprise-only/springboot/bus-app-security-disabling-proc.adoc]: "they will be unable to action group assigned tasks." Seems like it should say something like "they will be unable to assign actions to group assigned tasks."

File: doc-content/enterprise-only/springboot/bus-app-security-preauthenticate-proc.adoc [hunk 13]
```diff
@@ -0,0 +1,102 @@
+[id='bus-app-security-preauthenticate-proc_{context}']
+
+= Using Spring Security with preauthenication
+If you disable Spring Security authentication by using the `PermitAll` method, any user can log in to the application, but users will have limited access and functionality. However, you can preauthenticate a user, for example a designated service account, so a group of users can use the same login but have all of the permissions that they require. That way, you do not need to create credentials for each user.
+
```
Comment (michelehaglund) [inline doc-content/enterprise-only/springboot/bus-app-security-preauthenticate-proc.adoc:5]: Possible change to:

and have all of the permissions that they require without creating credentials for each user.

**Review by michelehaglund — APPROVED** [review]
I added a few comments.

File: doc-content/enterprise-only/springboot/bus-app-security-preauthenticate-proc.adoc [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+@Slf4j
```
Comment (afalhambra) [inline doc-content/enterprise-only/springboot/bus-app-security-preauthenticate-proc.adoc:35]: @emmurphy1 Please remove this line. Thanks!
```suggestion
```

**Review by afalhambra — CHANGES_REQUESTED** [review]

**Comment by emmurphy1** [comment]
@michelehaglund for some reason I am unable to comment on your comments, so here you go:
Add the following dependencies to the Spring Boot project's pom.xml file:
I'm trying to avoid anthropomorphism.

and have all of the permissions that they require without creating credentials for each user.
without requiring is vague (dangling participle, not sure. Anyway it's not clear who is responsible for creating (or not creating) the credentials

**Pull request closed without merging**